hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/PrngSystemContract.java

```diff
@@ -0,0 +1,138 @@
+import static com.hedera.node.app.service.contract.impl.utils.SystemContractUtils.contractFunctionResultFailedFor;
+import static com.hedera.node.app.service.contract.impl.utils.SystemContractUtils.contractFunctionResultSuccessFor;
+import static java.util.Objects.requireNonNull;
+import static org.hyperledger.besu.evm.frame.ExceptionalHaltReason.INVALID_OPERATION;
+
+import com.hedera.hapi.node.base.ContractID;
+import com.hedera.node.app.service.contract.impl.state.ProxyWorldUpdater;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.precompile.AbstractPrecompiledContract;
+import org.hyperledger.besu.evm.precompile.PrecompiledContract;
+
+/**
+ * System contract to generate random numbers. This will generate 256-bit pseudorandom number when
+ * no range is provided using n-3 record's running hash. If 32-bit integer "range" and 256-bit
+ * "seed" is provided returns a pseudorandom 32-bit integer X belonging to [0, range).
+ */
+@Singleton
+public class PrngSystemContract extends AbstractPrecompiledContract {
+    private static final String PRECOMPILE_NAME = "PRNG";
+    // random256BitGenerator(uint256)
+    static final int PSEUDORANDOM_SEED_GENERATOR_SELECTOR = 0xd83bf9a1;
+    public static final String PRNG_PRECOMPILE_ADDRESS = "0x169";
+    private long gasRequirement;
+
+    @Inject
+    public PrngSystemContract(@NonNull final GasCalculator gasCalculator) {
+        super(PRECOMPILE_NAME, gasCalculator);
+    }
+
+    @Override
+    public long gasRequirement(final Bytes bytes) {
+        return gasRequirement;
+    }
+
+    @Override
+    public @NonNull PrecompileContractResult computePrecompile(final Bytes input, @NonNull final MessageFrame frame) {
+        requireNonNull(input);
+        requireNonNull(frame);
+
+        // compute the gas requirement
+        gasRequirement =
+                calculateGas(Instant.ofEpochSecond(frame.getBlockValues().getTimestamp()));
+
+        // get the contract ID
+        final ContractID contractID = asEvmContractId(Address.fromHexString(PRNG_PRECOMPILE_ADDRESS));
+
+        try {
+            // compute the pseudorandom number
+            final var randomNum = generatePseudoRandomData(input, frame);
+            requireNonNull(randomNum);
+            final var result = PrecompiledContract.PrecompileContractResult.success(randomNum);
+
+            // create a child record if we are not in a static call
+            if (!frame.isStatic()) {
+                createSuccessfulRecord(frame, randomNum, contractID);
+            }
+
+            return result;
+        } catch (Exception e) {
```

Comment:
I can see at least two different reasons for this exception - invalid selector supplied, which is clear user error - and insufficient entropy because this is called before block 4 - which should never happen.  But they both show up the same way, is this correct?  (Maybe there's nothing else to do with a problem like invalid entropy that should never happen. But if it _does_ happen that's a problem, right?  Maybe ... log it?)

Reply:
There is a warn level log message in the mono service implementation.  I'll move it over.